src/ports/postgres/modules/utilities/pivot.py_in

```diff
@@ -58,66 +61,164 @@ def pivot(schema_madlib, source_table, out_table,
-    where the piv column has 3 distinct values (10.0, 20.0 and 30.0).
+    where the piv column has 3 distinct values (10, 20 and 30).
     If the pivot function call is :
         SELECT madlib.pivot('pivset', 'pivout', 'id', 'piv', 'val');
     We want to construct the following sql code to pivot the table.
         CREATE TABLE pivout AS (SELECT id,
-        sum(CASE WHEN "piv" = '10.0' THEN val ELSE NULL END ) as "piv_10.0",
-        sum(CASE WHEN "piv" = '20.0' THEN val ELSE NULL END ) as "piv_20.0",
-        sum(CASE WHEN "piv" = '30.0' THEN val ELSE NULL END ) as "piv_30.0"
+        avg(CASE WHEN "piv" = '10' THEN val ELSE NULL END ) as "val_avg_piv_10",
+        avg(CASE WHEN "piv" = '20' THEN val ELSE NULL END ) as "val_avg_piv_20",
+        avg(CASE WHEN "piv" = '30' THEN val ELSE NULL END ) as "val_avg_piv_30"
         FROM pivset GROUP BY id ORDER BY id)
 
     """
     indices = split_quoted_delimited_str(index)
-    pcol = split_quoted_delimited_str(pivot_cols)
-    pval = split_quoted_delimited_str(pivot_values)
-    validate_pivot_coding(source_table, out_table, indices, pcol, pval)
+    pcols = split_quoted_delimited_str(pivot_cols)
+    pvals = split_quoted_delimited_str(pivot_values)
+    validate_pivot_coding(source_table, out_table, indices, pcols, pvals)
+
+    # Strip end quotes from pivot columns
+    pcols = [strip_end_quotes(pcol.strip()) for pcol in pcols]
+
+    # Parse the aggregate_func as a dictionary
+    try:
+        param_types = dict.fromkeys(pvals, list)
+        param_types['__madlib_def_pval__'] = list
+        agg_dict = extract_keyvalue_params(aggregate_func,param_types)
+    except KeyError, e:
+        with MinWarning("warning"):
+            plpy.warning("Pivot: Not all columns from '{aggregate_func}' present"
+            " in '{pivot_values}'".format(**locals()))
+            raise
+
+    # If the dictionary is empty, parse it as a list
+    agg_set = split_quoted_delimited_str(aggregate_func) if len(agg_dict) < 1 \
+         else []
+    # If the list is empty set 'avg' as default
+    agg_dict['__madlib_def_agg__'] = ['avg'] if len(agg_set) < 1 else agg_set
+
+    # Find the distinct values of pivot_cols
+    # Note that the distinct values are appended in order of users list
+    # This ordering is important when we access pivot_comb entries
+    distinct_str = ["SELECT "]
+    distinct_str.append(
+        ', '.join("array_agg(DISTINCT {pcol}) AS {pcol}_values".
+            format(pcol=pcol) for pcol in pcols))
+    distinct_str.append(" FROM " + source_table)
+    distinct_values = plpy.execute(''.join(distinct_str))
+
+    # Collect the distinc values for every pivot column
+    pcol_distinct_values = []
+    for pcol in pcols:
+        # Read the distinct values for this pcol
+        pcol_tmp = [item for item in
+            distinct_values[0][pcol+"_values"]]
+        # Remove null values if keep null is not true
+        if not keep_null:
+            pcol_tmp = filter(None,pcol_tmp)
```

Comment:
This removes 0 values in addition to NULLs. I replaced it with `pcol_tmp = [x for x in pcol_tmp if x is not None]`